src/main.rs

```diff
@@ -150,37 +126,54 @@ impl<'a> Interpreter<'a> {
         }
     }
 
-    fn get_fn(&self, name: &str) -> &'a Func {
-        self.module.funcs.get(name).unwrap()
+    fn get_fn(&self, name: &str) -> Rc<Func> {
+        self.module.funcs[name].clone()
     }
 
-    fn trace(&mut self, o_func: &'a Func, o_pc: usize) -> (&'a Func, usize) {
+    // XXX: why do I return func, pc? shouldn't that be the same as the input?
+    fn trace(&mut self, instr: &InstrPtr) -> (InstrPtr, Trace) {
         use Instruction::*;
 
-        let mut pc = o_pc;
-        let mut func = o_func;
-
         let mut trace = Vec::new();
-        let mut stack_size = 0;
+
+        let mut call_tree = Stack::root(FrameInfo {
+            func: instr.func.clone(),
+            back_ref: self.frames.last().unwrap().back_ref.clone(),
+            offset: 0,
+        });
+
+        let mut locals = TraceDataAllocator::new();
+        locals.alloc(instr.func.args_count + instr.func.locals_count);
+
+        let mut next = instr.clone();
```

Comment:
Why do we have to `clone` instructions all the time? Is this a restriction imposed by the borrow checker or ... ?